apps/avifenc.c

```diff
@@ -195,7 +210,8 @@ static void syntaxLong(void)
     printf("                                        (use 2 for any you wish to leave unspecified)\n");
     printf("    -r,--range RANGE                  : YUV range [limited or l, full or f]. (JPEG/PNG only, default: full; For y4m or stdin, range is retained)\n");
     printf("    --target-size S                   : Set target file size in bytes (up to 7 times slower)\n");
-    printf("    --progressive                     : EXPERIMENTAL: Encode a progressive image\n");
+    printf("    --auto-progressive                : EXPERIMENTAL: Auto set parameters to encode a simple progressive image.\n");
+    printf("    --layer N                         : EXPERIMENTAL: Set number of layers for progressive image [1-%d]. (default: 1 means disable progressive)\n", AVIF_MAX_AV1_LAYER_COUNT);
```

Comment:
The reason why `--layer` is added, at the first place, is to support y4m input, in which number of frames is unknown beforehand. `--layer` being able to tweak `--auto-progressive` is just a good-to-have extra.

Reply:
> to support y4m input, in which number of frames is unknown beforehand

I would rather put the burden of counting frames on the avifenc implementation (even if it is a standalone function to do exclusively that) rather than on the user.

Reply:
> The issue is that for y4m you have to read the whole input to know how many frames are there in it, and sometimes you can only read the input once (the `stdin` case).

Yes, and I still would like avifenc to do that job for me.

> However given that we agreed to not support changing config halfway encoding a y4m input, progressive encoding with multi-frame y4m seems not very useful then, and we may just reject multi-frame y4m / discard remaining frames.

An error or a warning is fine with me. If and when that becomes a lacking feature, we will reconsider alternatives.

> Then we can use a boolean flag to enable progressive encoding.

Do you have a flag name in mind?

> But I prefer not merging this flag with `--auto-progressive` to avoid confusion. `--auto-progressive` tries to hide the differences between AVIF and other formats' progressive feature, but "manual progressive" requires understanding that AVIF's progressive feature is fundamentally different.

If novice users are targeted, I think keeping `--progressive` for the simplest use case is best. With a bit of `-h` reading, one can extend the flag use to "manual progressive" (more settings), so I do not find it that confusing. The only downside I see is the lack of error or warning when one tries to use `--progressive` with multiple frames and expecting an animation or something else than several layers.

Reply:
> Yes, and I still would like avifenc to do that job for me.

In case you didn't get it, I mean if we don't reject multi-frame Y4M inputs it's simply impossible for `avifenc` to do that. We now reject multi-frame Y4M inputs so now it's possible though.

> An error or a warning is fine with me.

Didn't border exhausting the y4m input manually or refactoring closing `y4mFrameIterator` as a dedicated function, so I simply made it an error.

> Do you have a flag name in mind?

I decided to go with `--layered`, and updated to call the feature "layered AVIF" instead of "progressive AVIF" in help string and comment (The name of this feature in AVIF standard is "AV1 Layered Image Items"). The word "progressive" now only referring to the rendering effect on supported viewers, or asking for AVIF with such effect (with single input, as is what users used to).

This should make it clear that:
- Layered AVIF supports progressive rendering.
- `--progressive` is asking for AVIF supporting progressive rendering. (Don't care what exact feature or config is used)
- `--layered` is asking for encoding a layered AVIF. (Intend to tweak the feature manually)

Reply:
> I mean if we don't reject multi-frame Y4M inputs it's simply impossible for `avifenc` to do that.

What is "that" exactly?  
If "that" is `avifenc` deducing the N from `--layer N` by itself, `avifenc` can just read the entire Y4M files to count the frames before doing anything else, no? I fail to see the impossibility, except for stdin input so large that it does not fit in memory. Maybe you meant without heavy avifenc.c refactoring?

> I decided to go with `--layered`

Good!

Reply:
> except for stdin input so large that it does not fit in memory.

Caching stdin in memory totally go against the whole point of piping, which is to avoid holding all in memory at once, so I was considering it a no-go.

Reply:
For large and lengthy video content, I agree. But for progressive still images, it sounds reasonable to me.